api/caseshare/real-api.ts

```diff
@@ -46,65 +48,111 @@ export async function getCases(req: EnhancedRequest, res: Response, next: NextFu
-      updatedErrorMessages.push(`{request: ${reason.config.data}, response: {${reason.data.status} ${reason.data.message}}}`)
+  shareCases.flatMap(sharedCase => {
+    if (sharedCase && sharedCase.pendingShares && sharedCase.pendingShares.length > 0) {
+      sharedCase.pendingShares.flatMap( pendingShare => {
+        const caseAssigneeMappingModel: CaseAssigneeMappingModel = {
+          assignee_id: pendingShare.idamId,
+          case_id: sharedCase.caseId,
+          case_type_id: sharedCase.caseTypeId,
+        }
+        const promise = sendPost(path, caseAssigneeMappingModel, req)
+        promises.push(promise)
+      })
     }
   })
-  const updatedSharedCases: SharedCase[] = handleRejectedPayloads(shareCases, rejectedPayloads)
-  return {
-    updatedErrorMessages,
-    updatedSharedCases,
+  return promises
+}
+
+function doUnShareCase(req: EnhancedRequest, shareCases: SharedCase[]): any {
+  const path = `${ccdUrl}/case-assignments`
+  const unassignedCaseModels = []
+  let promise = null
+  for (const aCase of shareCases) {
+    if (aCase.pendingUnshares && aCase.pendingUnshares.length > 0) {
+      const userDetails: UserDetails[] = aCase.pendingUnshares
+      for (const aUser of userDetails) {
+        const unassignedCaseModel: UnassignedCaseModel = {
+          assignee_id: aUser.idamId,
+          case_id: aCase.caseId,
+          case_roles: aUser.caseRoles,
+        }
+        unassignedCaseModels.push(unassignedCaseModel)
+      }
+    }
+  }
+  const payload = {
+    unassignments: unassignedCaseModels,
   }
+  if (unassignedCaseModels.length > 0) {
+    promise = sendDelete(path, payload, req)
+  }
+  return promise
 }
 
-function promiseCaseShareBatchCall(shareCases: SharedCase[], path: string, req: EnhancedRequest) {
-  // @ts-ignore
-  return shareCases.flatMap(sharedCase => sharedCase.pendingShares.flatMap( pendingShare => {
-    const caseAssigneeMappingModel: CaseAssigneeMappingModel = {
-      'assignee_id': pendingShare.idamId,
-      'case_id': sharedCase.caseId,
-      'case_type_id': sharedCase.caseTypeId,
+function handleRejectedResponse(allResults: any):
+  { updatedErrorMessages: string[],
+    rejectedShareCasePayloads: CaseAssigneeMappingModel[],
+    rejectedUnShareCaseReason: any,
+    rejectedCount: number} {
+  const updatedErrorMessages: string[] = []
+  const rejectedShareCasePayloads: CaseAssigneeMappingModel[] = []
+  let rejectedUnShareCaseReason: any = null
+  let rejectedCount: number = 0
+
+  allResults.forEach(result => {
+    const {status, reason}: {status: string, reason: any} = result
+    if (status === 'rejected') {
+      if (reason.config.method === 'post') {
+        rejectedShareCasePayloads.push(JSON.parse(reason.config.data))
+      } else if (reason.config.method === 'delete') {
+        rejectedUnShareCaseReason = reason
+      }
+      updatedErrorMessages.push(`{request: ${reason.config.data}, response: {${reason.data.status} ${reason.data.message}}}`)
```

Comment:
_Question:_
This looks like it will result in a peculiar not-quite-json structure:
`'{ request: {}, response: { 400 message } }'`

Reply:
Yes, as there are N api calls, we need to know which call for assign a user to a case have been failed.

Reply:
That wasn't what I was getting at - the string literal template produces something that _sort of_ looks like JSON but isn't. I don't know quite what you were going for. If it needs to be parseable as JSON, that will fail; if it doesn't, I'm not sure why it resembles JSON. 🤷‍♂️

Reply:
changed to something not JSON format.